File: AllToDo-Android/app/src/main/java/com/example/alltodo/wasm/WasmManager.kt
package com.example.alltodo.wasm

import android.content.Context
import com.google.gson.Gson
import com.example.alltodo.services.RemoteLogger
import okhttp3.OkHttpClient
import okhttp3.Request
import kotlinx.coroutines.*

import javax.inject.Inject
import javax.inject.Singleton
import dagger.hilt.android.qualifiers.ApplicationContext

@Singleton
class WasmManager @Inject constructor(
    @ApplicationContext private val context: Context
) {
    private val client = OkHttpClient()
    private val gson = Gson()
    private val storage = WasmStorage(context)
    private val runtime: WasmRuntime = WebViewWasmRuntime(context)
    private val advancedUrl = "http://175.194.163.56:8003/wasm/advanced"
    private val versionUrl = "http://175.194.163.56:8003/wasm/version"
    
    private val TAG = "[WASM_STATUS]"
    
    data class VersionResponse(val version: String)
    
    // [NEW] Debugging Properties for UI Overlay
    var lastErrorMessage: String? = null
        private set
        
    var onStatusUpdate: ((String) -> Unit)? = null

    fun initialize(onReady: (Boolean) -> Unit) {
        
        // 1. Initial Load from storage if available
        storage.load()?.let { (_, blobJson) ->
            try {
                val bundle = gson.fromJson(blobJson, WasmBundle::class.java)
                val decrypted = WasmCrypto.decrypt(bundle)
                runtime.loadModule(decrypted)
                RemoteLogger.info("WASM Init (Stored): Version ${bundle.version}")
                lastErrorMessage = null
                onStatusUpdate?.invoke("Ready (Stored v${bundle.version})")
            } catch (e: Exception) {
                e.printStackTrace()
                RemoteLogger.error("Failed to load stored WASM: ${e.message}")
                loadFallback()
            }
        } ?: run {
            RemoteLogger.info("No stored WASM. Using Fallback.")
            loadFallback()
            onStatusUpdate?.invoke("Ready (Fallback)")
        }

        // 2. Check for Updates in Background
        CoroutineScope(Dispatchers.IO).launch {
            checkForUpdate()
            onReady(true) // Notify valid state (either stored or fallback is ready)
        }
    }
    
    // [NEW] Exposed method to replace Native Code
    // [NEW] Exposed method to replace Native Code
    suspend fun compress(points: List<Int>): List<Int> {
        val start = System.currentTimeMillis()
        
        val result = runtime.compressTrajectory(points, 3, 10)
        
        val duration = System.currentTimeMillis() - start
        val msg = "WASM Success: ${points.size/2} -> ${result.size/2} pts (${duration}ms)"
        // RemoteLogger.info(msg)
        lastErrorMessage = null
        withContext(Dispatchers.Main) {
            onStatusUpdate?.invoke("Comp: ${points.size/2}->${result.size/2} (${duration}ms)")
        }
        return result
    }

    // [NEW] Clustering Support
    suspend fun cluster(points: List<Int>, cellSizeMeters: Int): List<Int> {
        val start = System.currentTimeMillis()
        
        val result = runtime.clusterPoints(points, cellSizeMeters)
        
        val duration = System.currentTimeMillis() - start
        lastErrorMessage = null
        // onStatusUpdate?.invoke("Cluster: ${duration}ms")
        return result
    }

    private suspend fun checkForUpdate() {
        try {
            // A. Check Server Version
            val request = Request.Builder().url(versionUrl).build()
            val response = client.newCall(request).execute()
            if (!response.isSuccessful) {
                return
            }
            
            val serverVersion = gson.fromJson(response.body?.string(), VersionResponse::class.java).version
            val (storedVersion, _) = storage.load() ?: ("0.0.0" to "")
            
            if (serverVersion != storedVersion) {
                if (fetchAndLoadAdvanced()) {
                    RemoteLogger.info("WASM Upgraded to Version $serverVersion")
                } else {
                }
            } else {
            }
        } catch (e: Exception) {
            e.printStackTrace()
        }
    }

    private suspend fun fetchAndLoadAdvanced(): Boolean {
        return try {
            val request = Request.Builder().url(advancedUrl).build()
            val response = client.newCall(request).execute()
            
            if (!response.isSuccessful) {
                return false
            }
            
            
            val body = response.body?.string() ?: return false
            val bundle = gson.fromJson(body, WasmBundle::class.java)
            val decrypted = WasmCrypto.decrypt(bundle)
            
            runtime.loadModule(decrypted)
            storage.save(bundle.version, body)
            
            
            // [NEW] Self-Test Routine
            verifyWasm()
            
            return true
        } catch (e: Exception) {
            e.printStackTrace()
            false
        }
    }

    private suspend fun verifyWasm() {
        try {
            // 1. RDP Test Case
            // (0,0) -> (1,1) -> (2,2) with thresh=5m. Expected: (0,0) -> (2,2)
            val rdpPoints = listOf(0, 0, 100000, 100000, 200000, 200000) 
            val rdpResult = runtime.compressTrajectory(rdpPoints, 5, 5)
            
            val rdpPassed = if (rdpResult.size == 4) {
                 val p1 = "${rdpResult[0]},${rdpResult[1]}"
                 val p2 = "${rdpResult[2]},${rdpResult[3]}"
                 (p1 == "0,0" && p2 == "200000,200000")
            } else false
            
            // 2. Clustering Test Case
            // 2 points close to each other: (0,0) and (10,10) (approx 1.4m dist)
            // Cell Size: 100m. Expected: 1 Cluster with count 2.
            val clusterPoints = listOf(0, 0, 100, 100)
            val clusterResult = runtime.clusterPoints(clusterPoints, 100)
            
            val clusterPassed = if (clusterResult.size == 3) {
                val count = clusterResult[2]
                count == 2
            } else false

            if (rdpPassed && clusterPassed) {
                onStatusUpdate?.invoke("WASM Verified (RDP+Cluster)")
            } else {
            }
        } catch (e: Exception) {
        }
    }

    private fun loadFallback() {
        try {
            val input = context.assets.open("fallback.wasm")
            val bytes = input.readBytes()
            runtime.loadModule(bytes)
            RemoteLogger.info("Loaded Fallback WASM")
            lastErrorMessage = null
        } catch (e: Exception) {
            e.printStackTrace()
            RemoteLogger.error("CRITICAL: Failed to load Fallback WASM")
            lastErrorMessage = "Failed to load WASM"
        }
    }
}

File: WasmProject/ios/WasmClient/WasmManager.swift
import Foundation

final class WasmManager {
    static let shared = WasmManager()
    private let runtime: WasmRuntime = DummyWasmRuntime()
    private let session = URLSession(configuration: .default)
    private let advancedURL = URL(string: "http://127.0.0.1:8000/wasm/advanced")!
    
    private init() {}
    
    func initialize(completion: @escaping (Bool) -> Void) {
        // 1. 저장된 WASM 먼저 사용 시도
        if let (version, blobJson) = WasmStorage.shared.load() {
            // Note: In storage usage, version is unused in this snippet but available.
            if let bundle = try? JSONDecoder().decode(WasmBundle.self, from: blobJson),
               let decrypted = try? WasmCrypto.shared.decrypt(bundle: bundle) {
                do {
                    try runtime.loadModule(decrypted)
                    completion(true)
                    return
                } catch {
                    print("Failed to load stored WASM: \(error)")
                }
            }
        }
        
        // 2. 서버에서 새로 다운로드 시도
        fetchAdvancedWasm { success in
            if success {
                completion(true)
            } else {
                // 3. 완전히 실패하면 fallback.wasm 사용
                self.loadFallback()
                completion(false)
            }
        }
    }
    
    private func fetchAdvancedWasm(completion: @escaping (Bool) -> Void) {
        session.dataTask(with: advancedURL) { data, _, error in
            guard let data = data, error == nil else {
                completion(false)
                return
            }
            
            do {
                let bundle = try JSONDecoder().decode(WasmBundle.self, from: data)
                let decrypted = try WasmCrypto.shared.decrypt(bundle: bundle)
                try self.runtime.loadModule(decrypted)
                
                // Save original JSON blob, not decrypted binary, as storage expects to decode Bundle again?
                // The snippet `WasmStorage.shared.save(version: bundle.version, blobJson: data)` saves the JSON response.
                // The load logic `JSONDecoder().decode...` confirms this.
                WasmStorage.shared.save(version: bundle.version, blobJson: data)
                
                completion(true)
            } catch {
                print("Failed to download/load advanced wasm: \(error)")
                completion(false)
            }
        }.resume()
    }
    
    private func loadFallback() {
        guard
            let url = Bundle.main.url(forResource: "fallback", withExtension: "wasm"),
            let data = try? Data(contentsOf: url)
        else {
            print("fallback.wasm not found")
            return
        }
        
        do {
            try runtime.loadModule(data)
            print("Loaded fallback WASM")
        } catch {
            print("Failed to load fallback WASM: \(error)")
        }
    }
}
